src/router.rs

```diff
@@ -1,127 +1,63 @@
-use crate::{alcazar::AlcazarError, http_request::HttpError};
+use crate::error::{AlcazarError, HttpError, Result};
+use crate::routing::endpoint::{Endpoint, MethodType};
 use std::str::FromStr;
 
-#[derive(PartialEq, Copy, Clone)]
-pub enum MethodType {
-    POST,
-    GET,
-    PATCH,
-    DELETE,
-    CONNECT,
-    OPTIONS,
-    TRACE,
-    HEAD,
-}
-
-impl FromStr for MethodType {
-    type Err = AlcazarError;
-
-    fn from_str(method: &str) -> Result<MethodType, AlcazarError> {
-        match method {
-            "POST" => Ok(MethodType::POST),
-            "GET" => Ok(MethodType::GET),
-            "PATCH" => Ok(MethodType::PATCH),
-            "DELETE" => Ok(MethodType::DELETE),
-            "CONNECT" => Ok(MethodType::CONNECT),
-            "OPTIONS" => Ok(MethodType::OPTIONS),
-            "TRACE" => Ok(MethodType::TRACE),
-            "HEAD" => Ok(MethodType::HEAD),
-            _ => Err(AlcazarError::HttpError(HttpError::MethodNotImplemented)),
-        }
-    }
-}
-
 #[derive(Clone)]
-pub struct Endpoint {
-    method: MethodType,
-}
-
-impl Default for Endpoint {
-    fn default() -> Self {
-        Self {
-            method: MethodType::GET,
-        }
-    }
-}
-
-impl Endpoint {
-    pub fn new() -> Self {
-        Endpoint::default()
-    }
-
-    pub fn set_method(mut self, method: MethodType) -> Self {
-        self.method = method;
-        self
-    }
-}
-
-#[derive(Clone)]
-pub struct Route {
-    path: String,
-    endpoint: Endpoint,
+pub struct Router {
+    endpoints: Vec<Endpoint>,
 }
 
-impl Default for Route {
+impl Default for Router {
     fn default() -> Self {
         Self {
-            path: "/".into(),
-            endpoint: Endpoint::default(),
+            endpoints: Vec::new(),
         }
     }
 }
 
-impl Route {
+impl Router {
     pub fn new() -> Self {
-        Route::default()
-    }
-
-    pub fn set_path(mut self, path: String) -> Self {
-        self.path = path;
-        self
-    }
-
-    pub fn set_endpoint(mut self, endpoint: Endpoint) -> Self {
-        self.endpoint = endpoint;
-        self
-    }
-
-    pub fn get_response(&self) -> &'static str {
-        "HTTP/1.1 200 OK\r\n\r\n"
+        Router::default()
     }
-}
-
-#[derive(Clone)]
-pub struct Router {
-    routes: Vec<Route>,
-}
 
-impl Default for Router {
-    fn default() -> Self {
-        Self { routes: Vec::new() }
+    // Returns a list of declared endpoints.
+    pub(crate) fn endpoints(&self) -> &Vec<Endpoint> {
+        &self.endpoints
     }
-}
 
-impl Router {
-    pub fn new() -> Self {
-        Router::default()
-    }
+    // TODO: Add handler parameter and set it up
+    pub fn with_endpoint(mut self, path: &str, methods: &[&str]) -> Self {
```

Comment:
With this implementation of the router, we are replacing the `Route` by the `Endpoint`, but in my mind the `Endpoint` was a part of a `Route`. 
The proposed implementation is right for me, but I just want to be sure that is not a mistake and we will drop the term of `Route` replaced by `Endpoint`. For that we have to keep in mind to add the handler and the namespace directly on the endpoint.

Reply:
Having the `Route` struct is excessive here by a couple of reasons:
- It creates more complex and verbose API for endusers
- Any endpoint is always associated with the specific route and no needed to create a separate entity for it 
- Adding a handler reference / used middlewares will be much essential to add as the part of the endpoint rather the route